### ISSUE TITLE
EVM: Different segmentation faults when running the test suite with EVMC

### ISSUE THREAD

**Issue by jlokier** [status-im/nimbus-eth1#575]
@mjfh reported that `make test ENABLE_EVMC=1` fails. From https://github.com/status-im/nimbus-eth1/pull/573#issuecomment-810999399 @mjfh:

- compiling as `make test ENABLE_EVMC=1` crashes at test no. 7 with

       [..]
       [OK] tests/fixtures/eth_tests/GeneralStateTests/stCreate2/returndatacopy_afterFailing_create.json
       Segmentation fault
       subtest no: 7 failed: testgeneralstatejson

       [Suite] tracer json tests
         [OK] tests/fixtures/TracerTests/block48915.json
         [OK] tests/fixtures/TracerTests/block46147.json
       [..]

  and can be reproduced as `./build/all_test 7`, and terminates at test no. 24 as

       [...]
       [OK] 30_keys_0_storage.json

       [Suite] Misc test suite
         [OK] EthAddress to int
       stack trace: (most recent call last)
       /status/nimbus-eth1/vendor/nimbus-build-system/vendor/Nim/lib/system/nimscript.nim(416, 18)
       /status/nimbus-eth1/nimbus.nimble(37, 8) testTask
       /status/nimbus-eth1/nimbus.nimble(34, 8) test
       /status/nimbus-eth1/vendor/nimbus-build-system/vendor/Nim/lib/system/nimscript.nim(260, 7) exec
       /status/nimbus-eth1/vendor/nimbus-build-system/vendor/Nim/lib/system/nimscript.nim(260, 7) Error: unhandled exception: FAILED: build/all_tests [OSError]
       make: *** [Makefile:112: test] Error 1

  and this one will not be reproduced running `./build/all_test 24`

Turns out it fails on different tests for different people, but not the targets in CI.  From https://github.com/status-im/nimbus-eth1/pull/573#issuecomment-811405065 @jlokier:

- No. 7 is fine for me on the master branch.
  However, no. 9 fails on master on Ubuntu Server 19.10 64-bit x86 with a SIGSEGV.
  On MacOS, all the tests pass fine.
  Reverting to before the last big submodule bump didn't change the result, so it's been around a while.

  The final error message is a rather cryptic `Error: unhandled exception: FAILED: build/all_tests [OSError]`, for which the cause is a SIGSEGV a few thousand lines earlier. I found it using `strace -f`.

  I wondered why it's not getting caught in CI, and saw we don't have a Linux 64-bit EVMC CI setup. Just Mac 64-bit (which works) and Linux 32-bit (which also works apparently).

**Comment by jlokier** [comment]
The "no 7" segfault occurs in "new generalstate json tests":

```
./build/all_tests 7
[Suite] new generalstate json tests
  [OK] tests/fixtures/eth_tests/GeneralStateTests/stSubroutine/simpleSubroutine.json
  [OK] tests/fixtures/eth_tests/GeneralStateTests/stSubroutine/subroutineAtEndOfCode.json
...
  [OK] tests/fixtures/eth_tests/GeneralStateTests/stReturnDataTest/returndatacopy_after_failing_delegatecall.json
  [OK] tests/fixtures/eth_tests/GeneralStateTests/stReturnDataTest/returndatacopy_afterFailing_create.json
  [OK] tests/fixtures/eth_tests/GeneralStateTests/stCreate2/returndatacopy_afterFailing_create.json
Segmentation fault
```
@mjfh, can you say what build environment this is occurring in, so we can reproduce it?

**Comment by jlokier** [comment]
The "no 9" segfault occurs in "persist block json tests":

```
./build/all_tests 9
[Suite] persist block json tests
  [OK] tests/fixtures/PersistBlockTests/block2463413.json
  [OK] tests/fixtures/PersistBlockTests/block1927662.json
  [OK] tests/fixtures/PersistBlockTests/block248032.json
  [OK] tests/fixtures/PersistBlockTests/block47205.json
  [OK] tests/fixtures/PersistBlockTests/block46400.json
  [OK] tests/fixtures/PersistBlockTests/block46402.json
Segmentation fault
```
This occurs consistently on a 64-bit x86 Ubuntu Server 19.10 system, using `make test EVMC_HOST=1`.  On this target the "no 7" segfault does not occur.

Confirmed with versions from b6ad47f3 (when `EVMC_HOST` was added) to cb957b71 (current master).

**Comment by jangko** [comment]
> saw we don't have a Linux 64-bit EVMC CI setup.

It has been commented out together with windows 64 bit EVMC CI test. Both linux 64bit and windows 64 bit test crash for different reason.

**Comment by mjfh** [comment]
the command

     ulimit -s 20000

does the job on my machine, tests run ok now. my original setting on deb/x64 was 8192, so this might explain the difference to the ubuntu machine. In the gdb i saw it crashing at a stack nesting level of  more than 6k.

**Comment by jangko** [comment]
that is a bad sign if we consume more stack than other programs do. too many layer calls. and it is not good at all if we have to tell our user, please set your system stack limit higher.
other eth client run the same tests, and they are not running out of stack.

I remember evmone use some kind of continuation call to reduce this layering.

**Comment by jlokier** [comment]
If it's just excessive stack use, that's ok, as we are intending to refactor out the deep recursion for many good reasons.  There's no need to worry about this error if that's all it is, let's just fix the recursion.  Maybe adjust the test environment temporarily to have a larger stack so other CI can continue.

**Comment by jlokier** [comment]
This issue is now resolved due to the three PRs shown above: #586, #588, #598.

Previous worst case stage usage for some EVM programs was about 13 MB, exceeding the default stack size on various targets.

Now it's down to about 550 kB worst case (EVMC mode with recursion), 50 kB (non-EVMC mode with recursion), and not much at all (non-EVMC with no recursion).  This is small enough to use in many threads.

It's been tested locally, but matrix CI results are taking a long time to even start, still queued after 3 hours.

Because this was an intermittent issue, only affecting some targets, and even then it wobbled about a bit, it's worth validating the fix now and into the future.  This has been done by limiting the stack to 1 MB on all targets during `make test`.
